[cli] Add `vc open` command

MatthewStanciu opened the pull request:
This PR adds a new command, `vc open`, which opens a few select pages on the dashboard.

<img width="555" alt="Screen Shot 2022-07-07 at 3 41 07 PM" src="https://user-images.githubusercontent.com/14811170/177883235-03161f11-10b9-45b7-b413-d75fa9596d8f.png">

You can also pass in a combination of the arguments `dash`, `latest`, and `prod` to skip the prompt and open the given page directly.

Examples:

- `vc open dash` opens the dashboard
- `vc open latest` opens the latest preview deployment
- `vc open dash latest` opens the latest preview deployment in the dashboard
- `vc open latest --prod`  opens the production deployment
- `vc open dash latest --prod`  opens the production deployment in the dashboard
- `vc open dash my-deployment-ji2fjij2.vercel.app` opens a specific deployment in the dashboard

### 📋 Checklist

<!--
  Please keep your PR as a Draft until the checklist is complete
-->

#### Tests

- [x] The code changed/added as part of this PR has been covered with tests
- [x] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

MatthewStanciu committed:
Add basic command

MatthewStanciu committed:
Add prod

MatthewStanciu committed:
Add args

MatthewStanciu committed:
Add help page

MatthewStanciu committed:
Add open to global help

MatthewStanciu committed:
Add tests

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

MatthewStanciu committed:
Track fixtures

MatthewStanciu committed:
Accidentally put gitignore in the ignored folder

MatthewStanciu committed:
Fix dumb typo

MatthewStanciu committed:
Small refactoring + show commands in list

MatthewStanciu committed:
Combine 2 similar tests into 1

MatthewStanciu committed:
Fix spacing in help

MatthewStanciu committed:
Add example to help

almonk commented:
❤️ this feature! I love the `vc open` menu command but I have some suggestions on how we could maybe make the other syntax more memorable below;

```
vc open dash # Open the dashboard
vc open latest # Open the latest preview deploymwnt
vc open dash latest # Open the latest preview deployment in the dashboard
vc open prod # Open the product deployment
vc open dash prod # Open the production deployment in the dashboard
```

I think this is a simple syntax and concept to remember, since now you only need to remember 3 keywords: `latest`, `prod`, `dash` and the combination of them allows you acess all routes.

FWIW I would also make sure we alias `production -> prod` and `dashboard -> dash` to avoid being annoying.

MatthewStanciu committed:
almonk suggestions

- Remove `--prod` and `inspect`
- Add `latest`, `prod`, and `[url]`

MatthewStanciu committed:
huh

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

MatthewStanciu committed:
`link-project` –> `ensure-link`

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

almonk commented on packages/cli/src/commands/deploy/args.ts:
```diff
@@ -21,6 +21,7 @@ export const help = () => `
       link                 [path]      Link local directory to a Vercel Project
       login                [email]     Logs into your account or creates a new one
       logout                           Logs out of your account
+      open                 [options]   Opens a project URL in your browser
```
Capitilisation: project -> Project

almonk reviewed (approved):
Some super minor copy stuff – but this is great

EndangeredMassa commented on packages/cli/test/unit/commands/open.test.ts:
```diff
@@ -0,0 +1,261 @@
+
+      useUser();
+      const team = useTeams('team_dummy');
+      teamSlug = team[0].slug;
```
It looks like other tests depend on this one running. I suggest avoiding test dependencies because it makes them brittle and impossible to run in isolation.

MatthewStanciu replied:
This was a good catch, thank you.

MatthewStanciu committed:
project –> Project

MatthewStanciu commented on packages/cli/src/commands/open.ts:
```diff
@@ -0,0 +1,300 @@
+
+  ${chalk.gray('–')} Open a specific deployment URL
+  
+    ${chalk.cyan(`$ ${getPkgName()} open [url]`)}
```
Nothing really, it's just like `open` except you can only open Vercel deployments. Happy to remove if it's confusing.

MatthewStanciu replied:
Just removed 👍

MatthewStanciu committed:
Fix tests depending on first one

MatthewStanciu committed:
Refactor

- Extract parts of the command into separate functions
- Defer requests until they are needed
- Small cleanup

EndangeredMassa committed:
Merge branch 'main' into add/vc-open

MatthewStanciu committed:
Remove more reundant 'not_found's

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

MatthewStanciu committed:
Alias prod –> production, dash –> dashboard

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

styfle commented on packages/cli/src/commands/open.ts:
```diff
@@ -0,0 +1,324 @@
+
+  ${chalk.gray('–')} Open the latest production deployment URL
+
+    ${chalk.cyan(`$ ${getPkgName()} open prod`)}
```
This should probably be `vc open latest --prod` to follow the same pattern as `vc deploy` and `vc deploy --prod`.

MatthewStanciu committed:
`prod` argument –> `--prod` flag

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

EndangeredMassa commented on packages/cli/src/commands/open.ts:
```diff
@@ -0,0 +1,325 @@
+
+    -h, --help                     Output usage information
+    --confirm                      Skip confirmation prompts
+    --prod                        Filter for production deployments
```
Is the alignment off here?

EndangeredMassa commented on packages/cli/src/commands/open.ts:
```diff
@@ -0,0 +1,325 @@
+    return 0;
+  }
+
+  if (!test) execa('open', [choice]);
```
This will only work on natively macOS, I believe. Should we use a package that will make sure this is cross-platform?

I've not used a package like this in the past. I found this one that looks alright: https://www.npmjs.com/package/open

MatthewStanciu replied:
Wow, thank you for catching this.

EndangeredMassa reviewed (changes requested):
I think we need to make sure this works cross-platform before merging

MatthewStanciu committed:
Fix alignment & use `open`

MatthewStanciu committed:
Remove "open a specific deployment url"

MatthewStanciu committed:
Remove test for no longer existing code

MatthewStanciu committed:
Merge branch 'main' into add/vc-open

dferber90 commented:
Today I found myself wanting this too (opened https://github.com/vercel/vercel/discussions/8174 before realizing this PR exists) 🥳 

I imagined `vc open` to open `vercel.com/:team/:project` right away, without any prompt. I can see myself being inconvenienced by the extra confirmation step or having to type the long form `vc open dash`.
